magicgui/type_map.py

```diff
@@ -369,6 +381,60 @@ def _deco(type_):
     return _deco if type_ is None else _deco(type_)
 
 
+@contextmanager
+def type_registered(
+    type_: _T,
+    *,
+    widget_type: WidgetRef | None = None,
+    return_callback: ReturnCallback | None = None,
+    **options,
+) -> Iterator[None]:
+    """Context manager that temporarily registers a widget type for a given `type_`.
+
+    When the context is exited, the previous widget type associations for `type_` is
+    restored.
+
+    Parameters
+    ----------
+    type_ : _T
+        The type for which a widget class or return callback will be provided.
+    widget_type : Optional[WidgetRef]
+        A widget class from the current backend that should be used whenever ``type_``
+        is used as the type annotation for an argument in a decorated function,
+        by default None
+    return_callback: Optional[callable]
+        If provided, whenever ``type_`` is declared as the return type of a decorated
+        function, ``return_callback(widget, value, return_type)`` will be called
+        whenever the decorated function is called... where ``widget`` is the Widget
+        instance, and ``value`` is the return value of the decorated function.
+    **options
+        key value pairs where the keys are valid `WidgetOptions`
+    """
+    tw = TypeWrapper(type_)
+    tw.resolve()
+    _type_ = tw.outer_type_
+
+    prev_type_def: Optional[WidgetTuple] = _TYPE_DEFS.get(_type_, None)
+    _type_ = register_type(
+        type_, widget_type=widget_type, return_callback=return_callback, **options
+    )
+    new_type_def: Optional[WidgetTuple] = _TYPE_DEFS.get(_type_, None)
+    try:
+        yield
+    finally:
+
+        if return_callback is not None:
+            _RETURN_CALLBACKS[_type_].remove(return_callback)
```

Comment:
in the previous implementation, you restore also value of `_RETURN_CALLBACKS` here you only remove some value. 

What happen if someone use `register_type` with both `widget_type` and `return_callback` inside? 

Maybe also callback should be removed only if is not present in `_RETURN_CALLBACKS[_type_]` before the context manager usage?

Reply:
> Maybe also callback should be removed only if is not present in _RETURN_CALLBACKS[_type_] before the context manager usage?

sounds good, added